### PR TITLE
Ready - Add MultiDelete Command, refactor some KV Command types

### PR DESCRIPTION
**Pull request by alexmoore**
Superscedes #487, and removes duplication between MultiFetch and MultiDelete.

### CODE REVIEW

File: src/main/java/com/basho/riak/client/api/commands/kv/FetchValue.java [hunk 1]
```diff
@@ -75,8 +75,8 @@
 {
 
     private final Location location;
-    private final Map<Option<?>, Object> options =
-            new HashMap<Option<?>, Object>();
+    private final Map<RiakOption<?>, Object> options =
```
Comment (lukebakken) [inline src/main/java/com/basho/riak/client/api/commands/kv/FetchValue.java:78]: Is anyone going to care that `Option` changed to `RiakOption`?

Also, `RiakOption` is in the `riak` package, making the `Riak` prefix redundant. Why the change?

File: src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+    @Override
+    protected RiakFuture<ResponseType, List<Location>> executeAsync(final RiakCluster cluster)
+    {
+        List<BaseCommand> operations = buildOperations();
+        MultiFuture future = new MultiFuture(locations);
+
+        Submitter submitter = new Submitter(operations, maxInFlight, cluster, future);
+
+        Thread t = new Thread(submitter);
+        t.setDaemon(true);
```
Comment (lukebakken) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java]: I modified the Python client to _not_ use daemon threads, since they could hide programming errors (not closing the client correctly, for instance). Should `setDaemon(false)` be used here?

File: src/main/java/com/basho/riak/client/api/commands/kv/MultiFetch.java [hunk 3]
```diff
@@ -57,15 +41,15 @@
...
- *     try
- *     {
- *          FetchValue.Response response = f.get();
- *          myResults.add(response.getValue(MyPojo.class));
- *     }
- *     catch (ExecutionException e)
- *     {
- *         // log error, etc.
- *     }
+ * try
```
Comment (lukebakken) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiFetch.java]: Probably should restore whitespace here.

**Comment by lukebakken** [comment]
Just a couple comments

File: src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+    @Override
+    protected RiakFuture<ResponseType, List<Location>> executeAsync(final RiakCluster cluster)
+    {
+        List<BaseCommand> operations = buildOperations();
+        MultiFuture future = new MultiFuture(locations);
+
+        Submitter submitter = new Submitter(operations, maxInFlight, cluster, future);
+
+        Thread t = new Thread(submitter);
```
Comment (srgg) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java]: Not sure, but it might be a good case to use Netty's Thread pool here

File: src/main/java/com/basho/riak/client/api/commands/kv/FetchValue.java (hunk 1 above)
Comment (alexmoore) [inline src/main/java/com/basho/riak/client/api/commands/kv/FetchValue.java:78]: TL;DR: It allows us to use either the FetchValue.Options or DeleteValue.options in the base MultiCommand class, for when we build the individual commands.  https://github.com/basho/riak-java-client/blob/788e0710a0d695701aba5bf63b229df657a7cbbe/src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java#L194-L206

I also had to switch FetchValue and DeleteValue's options to use the shared superclass `RiakOption` over the local/subclass `Option` type, but this just lets us reuse the options on the individual commands without casting:
https://github.com/basho/riak-java-client/blob/788e0710a0d695701aba5bf63b229df657a7cbbe/src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java#L70-L75

We can also then use it in the same way for the MultiCommand Builder: 
https://github.com/basho/riak-java-client/blob/788e0710a0d695701aba5bf63b229df657a7cbbe/src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java#L132-L132

Yay reuse!

> Also, RiakOption is in the riak package, making the Riak prefix redundant. Why the change?

RiakOption's always been named as such, no changes there.

File: src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java (hunk 2 above)
Comment (alexmoore) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java]: Maybe, it was that way previously, it was just refactored over to this new file.

File: src/main/java/com/basho/riak/client/api/commands/kv/MultiFetch.java (hunk 3 above)
Comment (alexmoore) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiFetch.java]: Bleh, good catch.  My reformatter likes to squish code examples.

File: src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java [hunk 5]
```diff
@@ -54,9 +54,8 @@ protected MultiCommand(Builder builder)
 
         Submitter submitter = new Submitter(operations, maxInFlight, cluster, future);
 
-        Thread t = new Thread(submitter);
-        t.setDaemon(true);
-        t.start();
+        new Thread(submitter).start();
+
```
Comment (lukebakken) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java:63]: If the threads that are created here aren't kept track of and `join()`-ed correctly, they could keep the entire JVM from shutting down cleanly.

I am assuming that is why `setDaemon(true)` was used previously - it's the lazy way of not having to deal with clean shutdown.

Is there a way that at the end of the `MultiCommand` these threads can be cleaned up? Does a `Submitter` know when all of its work is done and is able to exit and let the thread complete?

Comment (alexmoore) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java:63]: So the only thing the new thread is used for is kicking off the individual scalar commands, one at a time until it reaches it's max concurrency.  If we're interrupted while doing that we set `fail` on the multicommand future, and let the thread die.  This thread shouldn't have ownership over anything so letting it die is clean and humane. 

https://github.com/basho/riak-java-client/blob/17aa8296a7b642bd5b53737e37dcae28b54601d1/src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java#L253-L271

Once it reaches the end of the `run()` block, [the JVM will kill the thread](http://stackoverflow.com/questions/4283270/how-do-i-cleanup-a-thread-after-its-run-method-finishes), so no `join()` is needed. 

As for shutting it down... I'll have to think on that :) Maybe daemon is the way to go there.

Comment (lukebakken) [inline src/main/java/com/basho/riak/client/api/commands/kv/MultiCommand.java:63]: Right ... if there is a `close()` or `stop()` (`shutdown()`?) method on the client object, something in the `run()` block needs to know to make an early exit, and, most likely, the threads should be `join()`-ed so that client shutdown can be deterministic. May be more work than it's worth now, so a note could be made and `setDaemon(true)` put back in.

**Comment by lukebakken** [comment]
:shipit:

**Comment by alexmoore** [comment]
Thanks to @gerardstannard for the original PR! (#487).
